### DISCUSSION TITLE
Support lint rules in vue

### DISCUSSION THREAD

**Discussion by Boshen** [oxc-project/oxc#3370] [Feature Request]
https://github.com/vuejs/core/blob/main/eslint.config.js

**Comment by mysteryven** [comment]
Rules we need to implement:

- [x] [`eslint/no-restricted-globals`](https://eslint.org/docs/latest/rules/no-restricted-globals)
- [x] [`eslint/sort-imports`](https://eslint.org/docs/latest/rules/sort-imports#rule-details)
- [x] [`@typescript-eslint/consistent-type-imports`](https://typescript-eslint.io/rules/consistent-type-imports/)
	- [x] https://github.com/oxc-project/oxc/pull/3895
	- [x] https://github.com/oxc-project/oxc/pull/3984
- [x] [`@typescript-eslint/no-import-type-side-effects`](https://typescript-eslint.io/rules/no-import-type-side-effects/#:~:text=Enforce%20the%20use%20of%20top,specifiers%20with%20inline%20type%20qualifiers.&text=Some%20problems%20reported%20by%20this,fix%20ESLint%20command%20line%20option)
- [x] [`eslint-plugin-vitest/no-disabled-tests`](https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/no-disabled-tests.md) https://github.com/oxc-project/oxc/pull/3717
- [x] [`eslint-plugin-vitest/no-focused-tests`](https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/no-focused-tests.md) https://github.com/oxc-project/oxc/pull/4178
- [x] `eslint/no-restricted-syntax` in [vuejs/core](https://github.com/vuejs/core/blob/main/eslint.config.js#L28-L54) can be replaced with several rules

```js
{
	"oxc/no-const-enum": "error",
	"oxc/no-async-await": "error",
	"oxc/no-rest-spread-properties": ["error", {
		object_spread_message: "esbuild transpiles object spread into very verbose inline helpers.\n " + 
			"'Please use the `extend` helper from @vue/shared instead.",
		object_rest_message: "Our output target is ES2016, so async/await syntax should be avoided."
	}],
   "no-optional-chaining": [ "error", {
		message: "Our output target is ES2016, and optional chaining results in verbose helpers and should be avoided."
    }]
}
```
- [x] `import-x/no-nodejs-modules` can be replaced with [`prefer_node_protocol`](https://github.com/oxc-project/oxc/blob/c9d84af44e12735e98d59c12f9692a110cd8686d/crates/oxc_linter/src/rules/unicorn/prefer_node_protocol.rs)

```diff
-'import-x/no-nodejs-modules': [
-        'error',
-        { allow: builtinModules.map(mod => `node:${mod}`) },
-   ]
+ 'unicorn/prefer-node-protocol': 'error'
```

- [ ] [`no-unused-vars`](https://eslint.org/docs/latest/rules/no-unused-vars#rule-details) also needed, track it on #642.

**Comment by Boshen** [comment]
It seems like we need to support eslint v9 for per directory configuration :-/

```
files: ['packages/{vue,vue-compat,runtime-dom}/**'],
``` 

https://github.com/vuejs/core/blob/main/eslint.config.js

**Reply by mysteryven** [reply]
Or we can support `overrides` since our config file is v8.
- [`ESLint overrides`](https://eslint.org/docs/latest/use/configure/configuration-files-deprecated#how-do-overrides-work)
- [`Biome overrides`](https://biomejs.dev/reference/configuration/#overridesitemlinter)

ESLint following the rough steps:

1. Flatten config https://github.com/eslint/eslintrc/blob/59e890fcd9e03663ac185640d5bed5f1a85bd39b/lib/config-array-factory.js#L785-L791
2. Get final config for this file. https://github.com/eslint/eslint/blob/7bcda760369c44d0f1131fccaaf1ccfed5af85f1/lib/linter/linter.js#L1990
3. Get rules according config file https://github.com/eslint/eslint/blob/7bcda760369c44d0f1131fccaaf1ccfed5af85f1/lib/linter/linter.js#L1917

If we convert the config file to a group of thing, and can produce a final config for a specific path, seems like will not change too much for our current codes. But I don't know how much our performance be hurt, as need to generate new config file for each path.(I will take a look at the biome code) 

```rust
 pub fn run<'a>(&self, ctx: LintContext<'a>) -> Vec<Message<'a>> {
        let semantic = Rc::clone(ctx.semantic());

         // generate new eslint_config files and filter rules here.

		let ctx = ctx.with_fix(self.options.fix).with_eslint_config(&self.eslint_config);
        let rules = self
            .rules
            .iter()
            .map(|rule| {
                (rule, ctx.clone().with_rule_name(rule.name()).with_severity(rule.severity))
            })
            .collect::<Vec<_>>();

        for (rule, ctx) in &rules {
            rule.run_once(ctx);
        }
}
```

https://github.com/oxc-project/oxc/blob/3016f03578763379c4a1b86ed6d465dd42c5e8b7/crates/oxc_linter/src/lib.rs#L102-L103

**Comment by DonIsaac** [comment]
Ah, No unused vars, my arch enemy, maybe I'll try tackling that for the third time...

**Reply by Boshen** [reply]
It should be ready after all the recent changes. A simple search through all the unresolved references.

**Reply by DonIsaac** [reply]
The killer part is all the exceptions, particularly when allowing unused arguments that come before other used/ignored arguments...

Nonetheless I've got a v3 prototype up locally. I'll make a draft PR when it's ready.

**Comment by DonIsaac** [comment]
What's next up for this?

**Comment by drvn-mr** [comment]
Could `no-unused-vars` be supported in Vue `<script>` code?
I understand the doc mentions `.vue` files are entirely ignored for the rule since https://github.com/oxc-project/oxc/pull/4716

However `oxlint` does currently support the files (and JS/TS code) for other rules, excluding the template code.

I guess it was mentioned before: https://github.com/oxc-project/oxc/discussions/3370#discussioncomment-10109555

**Reply by Sysix** [reply]
We need to create an own template-parser to check if is used in vue. 
It was requested but not planned for the moment:

https://discord.com/channels/1079625926024900739/1271966101890142278/1272200920041521325

**Reply by fergusean** [reply]
@Sysix what about excluding no-unused-vars from processing top-level variables but still working within functions for variables that would never be exposed to the template due to scope?  I think there's still a lot of value in that, vs not working at all.
